src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -0,0 +1,1194 @@
+                bool isHeld = owningThreadId.IsInitialized && owningThreadId.Id == ThreadId.Current_NoInitialize.Id;
+                Debug.Assert(!isHeld || new State(this).IsLocked);
+                return isHeld;
+            }
+        }
+
+        internal static long ContentionCount => s_contentionCount;
+
+        internal void Dispose()
+        {
+            Debug.Assert(new State(this).IsInitialState);
+            _waitEvent?.Dispose();
+        }
+
+        internal nint LockIdForEvents
+        {
+            get
+            {
+                Debug.Assert(_waitEvent != null);
+                return _waitEvent.SafeWaitHandle.DangerousGetHandle();
+            }
+        }
+
+        internal unsafe nint ObjectIdForEvents
+        {
+            get
+            {
+                Lock lockObj = this;
+                return *(nint*)Unsafe.AsPointer(ref lockObj);
+            }
+        }
+
+        internal ulong OwningThreadId => _owningThreadId;
+
+        private static int DetermineMaxSpinCount() =>
+            AppContextConfigHelper.GetInt32Config(
+                "System.Threading.Lock.SpinCount",
+                "DOTNET_Lock_SpinCount",
+                DefaultMaxSpinCount,
+                allowNegative: false);
+
+        private static int DetermineMinSpinCount()
+        {
+            int adaptiveSpinPeriod =
+                AppContextConfigHelper.GetInt32Config(
+                    "System.Threading.Lock.AdaptiveSpinPeriod",
+                    "DOTNET_Lock_AdaptiveSpinPeriod",
+                    DefaultAdaptiveSpinPeriod,
+                    allowNegative: true);
+            if (adaptiveSpinPeriod < -1)
+            {
+                adaptiveSpinPeriod = DefaultAdaptiveSpinPeriod;
+            }
+
+            return -adaptiveSpinPeriod;
+        }
+
+        private struct State : IEquatable<State>
+        {
+            // Layout constants for _state
+            private const uint IsLockedMask = (uint)1 << 0; // bit 0
+            private const uint ShouldNotPreemptWaitersMask = (uint)1 << 1; // bit 1
+            private const uint SpinnerCountIncrement = (uint)1 << 2; // bits 2-4
+            private const uint SpinnerCountMask = (uint)0x7 << 2;
+            private const uint IsWaiterSignaledToWakeMask = (uint)1 << 5; // bit 5
+            private const byte WaiterCountShift = 6;
+            private const uint WaiterCountIncrement = (uint)1 << WaiterCountShift; // bits 6-31
+
+            private uint _state;
+
+            public State(Lock lockObj) : this(lockObj._state) { }
+            private State(uint state) => _state = state;
+
+            public static uint InitialStateValue => 0;
+            public static uint LockedStateValue => IsLockedMask;
+            private static uint Neg(uint state) => (uint)-(int)state;
+            public bool IsInitialState => this == default;
+            public bool IsLocked => (_state & IsLockedMask) != 0;
```

Comment:
```suggestion
            public readonly bool IsInitialState => this == default;
            public readonly bool IsLocked => (_state & IsLockedMask) != 0;
```
Or is there no (real) need for `readonly` in these cases?
I'll stop commenting on these places in order to don't make too much noise.